### PR TITLE
karma to jest migration completed

### PR DESCRIPTION
**Pull request by polbene95**
Closes https://github.com/Hacker0x01/react-datepicker/issues/4107

In this PR we are migrating from Karma to Jest.

Extra changes: 
-  In `day.jsx`, `month.jsx`, & `week.jsx` `PropTypes.instanceOf(Element)` was changed for `PropTypes.object` because the first one doesn’t work on server side render making the test fail.

- In `day.jsx` we changed `shouldFocusDay && this.dayEl.current.focus({ preventScroll: true });` for `shouldFocusDay && this.dayEl.current?.focus({ preventScroll: true });` because in some test `this.dayEl.current` was `null` making the test fail.

What's missing:

- Scripts for `test`, `test:ci` & `test:watch` need to be updated.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -90,15 +84,15 @@
     "rollup-plugin-replace": "^2.2.0",
     "rollup-plugin-terser": "^7.0.2",
     "rollup-plugin-uglify": "^6.0.4",
-    "sass": "1.66.1",
```
Comment (martijnrusschen) [inline package.json:93]: Why is this needed?

**Comment by martijnrusschen** [comment]
Looking good overall, would be good to get CI scripts running so we can have the test suite running.

File: package.json [hunk 2]
```diff
@@ -63,17 +65,9 @@
     "highlight.js": "^11.0.1",
     "husky": "8",
     "istanbul-instrumenter-loader": "^3.0.1",
-    "karma": "^6.4.2",
-    "karma-chai": "^0.1.0",
-    "karma-coverage": "^2.2.1",
```
Comment (martijnrusschen) [inline package.json:68]: We should consider finding an alternative for this. I think Jest may have support for this, but not sure how to set this up.

File: package.json [hunk 3]
```diff
@@ -124,7 +118,7 @@
     "lint": "yarn run eslint && yarn run sass-lint",
     "prettier": "prettier --write '**/*.{js,jsx}'",
     "start": "yarn --cwd docs-site install && yarn --cwd docs-site start",
-    "test": "NODE_OPTIONS=--openssl-legacy-provider NODE_ENV=test karma start karma.conf.js --single-run",
+    "test": "jest --verbose",
     "test:ci": "NODE_ENV=test karma start karma.conf.js --single-run",
     "test:watch": "NODE_OPTIONS=--openssl-legacy-provider NODE_ENV=test karma start karma.conf.js --watch",
```
Comment (martijnrusschen) [inline package.json:123]: These 2 lines need to be updated as well.

File: package.json (hunk 1 above)
Comment (polbene95) [reply package.json:93]: I don't know how did this got here. I reverted the change.

File: package.json (hunk 3 above)
Comment (polbene95) [reply package.json]: Just updated the scripts.

File: package.json (hunk 2 above)
Comment (polbene95) [reply package.json:68]: What do we need to find an alternative for exactly?

Comment (martijnrusschen) [reply package.json:68]: The code coverage tool

**Review by Zarthus — APPROVED** [review]
Thanks so much for doing this!

I reviewed all the non-test files and it gets a big thumbs up from me.

I skimmed over the new tests and they look good to me as well.
